### PR TITLE
Export grants to s3

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## 🎫 Ticket
BAU

## 📝 Description
Allow exporting grants/reports from a deployed environment to s3, anonymously, so that we can import their config locally for review and testing.

Corresponding terraform PR: https://github.com/communitiesuk/funding-service-terraform/pull/93

I've deployed this to dev and run an export successfully.

### Review Readiness
- [x] PR title and description provide sufficient context for reviewers
- [x] Commits are logical, self-contained, and have clear descriptions

### Testing
- [x] I have tested this change and it meets the acceptance criteria for the ticket
- [x] I need the reviewer(s) to pull and run this change locally
- [x] New (non-developer) functionality has appropriate unit and integration tests
- [-] End-to-end tests have been updated (if applicable)
- [-] Edge cases and error conditions are tested

### CODE REVIEW

File: app/developers/commands.py [hunk 1]
```diff
@@ -248,31 +271,45 @@ def export_grants(grant_ids: list[uuid.UUID], output: str) -> None:  # noqa: C90
...
-            first_name = faker.first_name()
-            last_name = faker.last_name()
-            user_data["email"] = f"{first_name.lower()}.{last_name.lower()}@test.communities.gov.uk"
-            user_data["name"] = f"{first_name} {last_name}"
+        for user in users:
+            export_data = __replace_id(export_data, str(user.id), str(placeholder_user.id))
 
-        export_data["users"].append(user_data)
+        export_data["users"] = [to_dict(placeholder_user)]
+        export_data["user_roles"] = []
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/developers/commands.py:284]: not bothered about performance here

File: app/developers/commands.py [hunk 2]
```diff
@@ -292,9 +328,23 @@ def export_grants(grant_ids: list[uuid.UUID], output: str) -> None:  # noqa: C90
...
+            )
+            s3_service.upload_file(buf, key=s3_key)
+            bucket = current_app.config["AWS_S3_BUCKET_NAME"]
+            click.echo(f"Written {len(grants)} grants to s3://{bucket}/{s3_key}")
+
 
 @developers_blueprint.cli.command("seed-grants", help="Load exported grants into the database")
 def seed_grants() -> None:  # noqa: C901
+    if current_app.config["IS_PRODUCTION"]:
+        raise click.ClickException("seed-grants must not be run in production; it deletes and recreates grants.")
```
Comment (sfount) [inline app/developers/commands.py:346]: Feels like a helpful check 👌

File: app/developers/commands.py [hunk 3]
```diff
@@ -173,17 +174,38 @@ def _import_organisations_and_handle_org_ids(export_data: ExportData) -> ExportD
...
+        exclude_users = True
+    else:
+        exclude_users = bool(exclude_users)
+
+    if output == "s3":
+        if not s3_key:
+            raise click.ClickException("--s3-key is required when --output=s3")
+        required_prefix = current_app.config["GRANT_EXPORT_FILES_PREFIX"].rstrip("/") + "/"
+        if not s3_key.startswith(required_prefix):
+            raise click.ClickException(f"--s3-key must start with {required_prefix!r}")
```
Comment (sfount) [inline app/developers/commands.py:201]: Today I learned about interpolating with `!r`

File: app/developers/commands.py [hunk 4]
```diff
@@ -248,31 +270,45 @@ def export_grants(grant_ids: list[uuid.UUID], output: str) -> None:  # noqa: C90
...
-        # Anonymise the user, but in a consistent way
-        if user.email not in USERS_NOT_TO_ANONYMISE:
-            faker = Faker()
-            faker.seed_instance(int(hashlib.md5(str(user_data["id"]).encode()).hexdigest(), 16))
-            first_name = faker.first_name()
-            last_name = faker.last_name()
-            user_data["email"] = f"{first_name.lower()}.{last_name.lower()}@test.communities.gov.uk"
-            user_data["name"] = f"{first_name} {last_name}"
+        for user in users:
+            export_data = __replace_id(export_data, str(user.id), str(placeholder_user.id))
```
Comment (sfount) [inline app/developers/commands.py:281]: Just to sense check this part anonymises everywhere a "created_by" or similar points to a user ID, which then the whole list of users gets replaced with the one anonymous user

**Review by sfount — APPROVED** [review]
All looks good to me, I did wonder if this script should just focus on exporting to stdout and then have the ad-hoc script calling it upload to s3 but given the number of changes it will have to make to behaviour anyway and the fact it already has stdout or file out options this interface all feels nice

File: app/developers/commands.py (hunk 4 above)
Comment (samuelhwilliams) [reply app/developers/commands.py:281]: yeah that's right. It's more about reducing the number of users that get exported than it is about anonymisation - the standard script always makes an attempt to anonymise personal details on user models. But for exporting grant/report config from prod, we don't care about who did what and there's potential for a lot of different created bys
